tools/build_bucket_golden_scraper/lib/build_bucket_golden_scraper.dart

```diff
@@ -0,0 +1,205 @@
+final class BuildBucketGoldenScraper {
+  /// Creates a scraper with the given configuration.
+  BuildBucketGoldenScraper({
+    required this.pathOrUrl,
+    this.dryRun = false,
+    String? engineSrcPath,
+    StringSink? outSink,
+  }) :
+      engine = engineSrcPath != null ?
+          Engine.fromSrcPath(engineSrcPath) :
+          Engine.findWithin(p.dirname(p.fromUri(io.Platform.script))),
+      _outSink = outSink ?? io.stdout;
+
+  /// Creates a scraper from the command line arguments.
+  ///
+  /// Throws [FormatException] if the arguments are invalid.
+  factory BuildBucketGoldenScraper.fromCommandLine(
+    List<String> args, {
+    StringSink? outSink,
+    StringSink? errSink,
+  }) {
+    outSink ??= io.stdout;
+    errSink ??= io.stderr;
+
+    final ArgResults argResults = _argParser.parse(args);
+    if (argResults['help'] as bool) {
+      _usage(args);
+    }
+    final String? pathOrUrl = argResults.rest.isEmpty ? null : argResults.rest.first;
+    if (pathOrUrl == null) {
+      _usage(args);
+    }
+    return BuildBucketGoldenScraper(
+      pathOrUrl: pathOrUrl,
+      dryRun: argResults['dry-run'] as bool,
+      outSink: outSink,
+      engineSrcPath: argResults['engine-src-path'] as String?,
+    );
+  }
+
+  static Never _usage(List<String> args) {
+    final StringBuffer output = StringBuffer();
+    output.writeln('Usage: build_bucket_golden_scraper [options] <path or URL>');
+    output.writeln();
+    output.writeln(_argParser.usage);
+    throw FormatException(output.toString(), args.join(' '));
+  }
+
+  static final ArgParser _argParser = ArgParser()
+    ..addFlag(
+      'help',
+      abbr: 'h',
+      help: 'Print this help message.',
+      negatable: false,
+    )
+    ..addFlag(
+      'dry-run',
+      help: "If true, don't write any files to disk (other than temporary files).",
+      negatable: false,
+    )
+    ..addOption(
+      'engine-src-path',
+      help: 'The path to the engine source code.',
+      valueHelp: 'path/that/contains/src (defaults to the directory containing this script)',
+    );
+
+  /// A local path or a URL to a buildbucket log file.
+  final String pathOrUrl;
+
+  /// If true, don't write any files to disk (other than temporary files).
+  final bool dryRun;
+
+  /// The path to the engine source code.
+  final Engine engine;
+
+  /// How to print output, typically [io.stdout].
+  final StringSink _outSink;
+
+  /// Runs the scraper.
+  Future<int> run() async {
+    final io.File file;
+
+    // If the path is a URL, download it and store it in a temporary file.
+    final Uri? maybeUri = Uri.tryParse(pathOrUrl);
+    if (maybeUri == null) {
+      throw FormatException('Invalid path or URL: $pathOrUrl');
+    }
+    if (maybeUri.hasScheme) {
+      file = await _downloadFile(maybeUri);
+    } else {
+      file = io.File(pathOrUrl);
+    }
+
+    // Read the file as a string.
+    final String contents = await file.readAsString();
+
```

Comment:
Can the file and temp directory be cleaned up after this point? I guess if you read the contents in `_downloadFile()` you can do all the cleanup there and make sure the temp dir gets cleaned up if the http client throws an exception.

Reply:
I can do you one better, why am I creating a temp file at all? :P